psh/hm/hm.c

```diff
@@ -121,20 +126,149 @@ void psh_hminfo(void)
 }
 
 
+static int psh_hm_cmppid(const void *t1, const void *t2)
+{
+	return (int)((threadinfo_t *)t1)->pid - (int)((threadinfo_t *)t2)->pid;
+}
+
+
+static void psh_hm_freeProcesses(rbtree_t *ptree)
+{
+	rbnode_t *n;
+	proc_t *p;
+
+	for (n = ptree->root; n != NULL; n = ptree->root) {
+		p = lib_treeof(proc_t, node, n);
+		lib_rbRemove(ptree, n);
+		free(p);
+	}
+}
+
+
+static int psh_hm_getProcesses(rbtree_t *ptree)
+{
+	int tcnt, i, j, n = 32;
+	threadinfo_t *info, *rinfo;
+	proc_t *p;
+
+	if ((info = malloc(n * sizeof(threadinfo_t))) == NULL) {
+		fprintf(stderr, "hm: Out of memory\n");
+		return -ENOMEM;
+	}
+
+	while ((tcnt = threadsinfo(n, info)) >= n) {
+		n *= 2;
+		if ((rinfo = realloc(info, n * sizeof(threadinfo_t))) == NULL) {
+			fprintf(stderr, "hm: Out of memory\n");
+			free(info);
+			return -ENOMEM;
+		}
+		info = rinfo;
+	}
+
+	qsort(info, tcnt, sizeof(threadinfo_t), psh_hm_cmppid);
+
+	i = 0;
+
+	while (i < tcnt) {
+		p = calloc(1, sizeof(*p));
+		if (p == NULL) {
+			fprintf(stderr, "hm: Out of memory\n");
+			psh_hm_freeProcesses(ptree);
+			return -ENOMEM;
+		}
+		p->pid = info[i].pid;
+		lib_rbInsert(ptree, &p->node);
+
+		for (j = i + 1; j < tcnt && info[j].pid == info[i].pid; j++)
+			;
+
+		i = j;
+	}
+
+	free(info);
+
+	return EOK;
+}
+
+
 int psh_hm(int argc, char *argv[])
 {
-	int i, err, status, progs = 0;
+	int c, i, err, status, check = 0, parent = 0, interval = 300, progs = 0;
 	proc_t *p, t;
-	pid_t pid;
+	pid_t pid, ppid;
+	rbtree_t ptree;
+	rbnode_t *n;
 
 	if (argc <= 1) {
 		psh_hm_help();
 		return EXIT_FAILURE;
 	}
 
+	while ((c = getopt(argc, argv, "cpt:h")) != -1) {
+		switch (c) {
+			case 'c':
+				check = 1;
+				break;
+
+			case 'p':
+				parent = 1;
+				break;
+
+			case 't':
+				interval = atoi(optarg);
+				break;
+
+			case 'h':
+				psh_hm_help();
+				return EOK;
+
+			default:
+				return EOK;
+		}
+	}
+
 	lib_rbInit(&hm_common.ptree, psh_hm_ptreeCompare, NULL);
 
-	for (i = 1; i < argc; ++i) {
+	if (check) {
```

Comment:
Maybe put this in separate function? This new feature does not in any way work together with the old one - it might be a better idea to do new applet with this, but let's at least separate this features.